src/openrct2/interface/Window.cpp

```diff
@@ -205,76 +210,40 @@ void WindowSetWindowLimit(int32_t value)
  */
 void WindowClose(WindowBase& w)
 {
-    auto itWindow = WindowGetIterator(&w);
-    if (itWindow == g_window_list.end())
-        return;
-
-    // Explicit copy of the shared ptr to keep the memory valid.
-    std::shared_ptr<WindowBase> window = *itWindow;
-
-    WindowEventCloseCall(window.get());
+    WindowEventCloseCall(&w);
 
     // Remove viewport
-    window->RemoveViewport();
+    w.RemoveViewport();
 
     // Invalidate the window (area)
-    window->Invalidate();
+    w.Invalidate();
 
-    // The window list may have been modified in the close event
-    itWindow = WindowGetIterator(&w);
-    if (itWindow != g_window_list.end())
-        g_window_list.erase(itWindow);
+    w.flags |= WF_DEAD;
+}
+
+void WindowFlushDead()
+{
+    // Remove all windows in g_window_list that have the WF_DEAD flag
+    g_window_list.remove_if([](const std::shared_ptr<WindowBase>& w) -> bool { return w->flags & WF_DEAD; });
 }
 
 template<typename TPred> static void WindowCloseByCondition(TPred pred, uint32_t flags = WindowCloseFlags::None)
 {
-    bool listUpdated;
-    do
+    for (auto it = g_window_list.rbegin(); it != g_window_list.rend(); ++it)
```

Comment:
What about the delete in reverse code

Reply:
Iterating in reverse is now the default so when it wants to close the top most window it just breaks out once it found it, this simplified the whole thing.

Reply:
And what about when a window opens a window during a close all

Reply:
It will just create it and marks the one window dead which will be erased later.

Reply:
My concern is you delete all windows when switching scenes but during the delete a new window is created which isn't deleted. This created window is then visible in the new scene.

Reply:
The actual delete won't cause new windows to be created, this would happen in OnClose before the window is actually freed and the deleted windows will be simply ignored in case it runs into an iteration before they are deleted. I've done quite rigorous testing and also looked at older issues/prs to make sure none of those problems are in there, the most prominent one was the scenario editor which I also tested under debug builds and everything works as expected.